tempesta_fw/procfs.c

```diff
@@ -0,0 +1,180 @@
+#include <linux/proc_fs.h>
+#include <linux/seq_file.h>
+
+#include "procfs.h"
+
+DEFINE_PER_CPU_ALIGNED(tfw_perfstat_t, tfw_perfstat);
+
+static struct proc_dir_entry *tfw_procfs_tempesta;
+static struct proc_dir_entry *tfw_procfs_perfstat;
+
+void
+tfw_perfstat_collect(tfw_perfstat_t *stat)
+{
+	int cpu;
+
+	preempt_disable();
+	for_each_possible_cpu(cpu) {
+		tfw_perfstat_t *pcpu_stat = per_cpu_ptr(&tfw_perfstat, cpu);
+
+		/* Client related statistics. */
+		stat->clnt.msgs_received += pcpu_stat->clnt.msgs_received;
+		stat->clnt.msgs_forwarded += pcpu_stat->clnt.msgs_forwarded;
+		stat->clnt.msgs_notparsed += pcpu_stat->clnt.msgs_notparsed;
+		stat->clnt.msgs_filterout += pcpu_stat->clnt.msgs_filterout;
+		stat->clnt.msgs_otherrors += pcpu_stat->clnt.msgs_otherrors;
+
+		stat->clnt.connects += pcpu_stat->clnt.connects;
+		stat->clnt.disconnects += pcpu_stat->clnt.disconnects;
+		stat->clnt.connections += pcpu_stat->clnt.connections;
+
+		stat->clnt.bytes_received += pcpu_stat->clnt.bytes_received;
+
+		/* Server related statistics. */
+		stat->serv.msgs_received += pcpu_stat->serv.msgs_received;
+		stat->serv.msgs_forwarded += pcpu_stat->serv.msgs_forwarded;
+		stat->serv.msgs_notparsed += pcpu_stat->serv.msgs_notparsed;
+		stat->serv.msgs_filterout += pcpu_stat->serv.msgs_filterout;
+		stat->serv.msgs_otherrors += pcpu_stat->serv.msgs_otherrors;
+
+		stat->serv.connects += pcpu_stat->serv.connects;
+		stat->serv.disconnects += pcpu_stat->serv.disconnects;
+		stat->serv.connections += pcpu_stat->serv.connections;
+
+		stat->serv.bytes_received += pcpu_stat->serv.bytes_received;
+	}
+	preempt_enable();
+}
+
+static int
+tfw_perfstat_seq_show(struct seq_file *seq, void *off)
+{
+	int ret;
+	tfw_perfstat_t stat;
+
+	memset(&stat, 0, sizeof(stat));
+	tfw_perfstat_collect(&stat);
+
+	/* Client related statistics. */
+	ret = seq_puts(seq, "== Tempesta FW client statistics ==\n");
```

Comment:
There is no need for the header line. Just see how statistic files like ```/proc/slabinfo``` or ```/proc/meminfo``` are formed.